velox/connectors/hive/iceberg/IcebergDataSink.cpp

```diff
@@ -0,0 +1,158 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/iceberg/IcebergDataSink.h"
+
+#include "velox/common/base/Fs.h"
+#include "velox/connectors/hive/TableHandle.h"
+
+namespace facebook::velox::connector::hive::iceberg {
+
+namespace {
+#define WRITER_NON_RECLAIMABLE_SECTION_GUARD(index)       \
+  memory::NonReclaimableSectionGuard nonReclaimableGuard( \
+      writerInfo_[(index)]->nonReclaimableSectionHolder.get())
+} // namespace
+
+IcebergDataSink::IcebergDataSink(
+    facebook::velox::RowTypePtr inputType,
+    std::shared_ptr<const IcebergInsertTableHandle> insertTableHandle,
+    const facebook::velox::connector::ConnectorQueryCtx* connectorQueryCtx,
+    facebook::velox::connector::CommitStrategy commitStrategy,
+    const std::shared_ptr<const HiveConfig>& hiveConfig)
+    : HiveDataSink(
+          std::move(inputType),
+          std::move(insertTableHandle),
+          connectorQueryCtx,
+          commitStrategy,
+          hiveConfig) {}
+
+std::vector<std::string> IcebergDataSink::close() {
+  checkRunning();
+  state_ = State::kClosed;
+  closeInternal();
+
+  auto icebergInsertTableHandle =
+      std::dynamic_pointer_cast<const IcebergInsertTableHandle>(
+          insertTableHandle_);
+
+  std::vector<std::string> commitTasks;
+  commitTasks.reserve(writerInfo_.size());
+  auto fileFormat = toString(insertTableHandle_->storageFormat());
+  std::string finalFileFormat(fileFormat);
+  std::transform(
+      finalFileFormat.begin(),
+      finalFileFormat.end(),
+      finalFileFormat.begin(),
+      ::toupper);
+
+  for (int i = 0; i < writerInfo_.size(); ++i) {
+    const auto& info = writerInfo_.at(i);
+    VELOX_CHECK_NOT_NULL(info);
+    // TODO(imjalpreet): Collect Iceberg file format metrics required by
+    // com.facebook.presto.iceberg.MetricsWrapper->org.apache.iceberg.Metrics#Metrics
+    // clang-format off
+      auto commitDataJson = folly::toJson(
+       folly::dynamic::object
+          ("path", info->writerParameters.writeDirectory() + "/" + info->writerParameters.writeFileName())
+          ("fileSizeInBytes", ioStats_.at(i)->rawBytesWritten())
+          ("metrics", folly::dynamic::object
+            ("recordCount", info->numWrittenRows))
+          ("partitionSpecId", icebergInsertTableHandle->partitionSpec()->specId)
+          ("partitionDataJson", partitionData_[i]->toJson())
+          ("fileFormat", finalFileFormat)
+          ("referencedDataFile", nullptr));
+    // clang-format on
+    commitTasks.push_back(commitDataJson);
+  }
+  return commitTasks;
+}
+
+void IcebergDataSink::splitInputRowsAndEnsureWriters(RowVectorPtr input) {
+  VELOX_CHECK(isPartitioned());
+
+  std::fill(partitionSizes_.begin(), partitionSizes_.end(), 0);
+
+  const auto numRows = partitionIds_.size();
+  for (auto row = 0; row < numRows; ++row) {
+    auto id = getWriterId(row);
```

Comment:
Line 91-105 seems to be the same as what's in HiveDataSink. Maybe extract a common function

Reply:
Refactored, thanks.